[feature] final vite migration

shellyear opened the pull request.

blcham commented on vite.config.js:
```diff
@@ -0,0 +1,26 @@
+    emptyOutDir: true,
+  },
+  define: {
+    'process.env': process.env
```
Not sure, but i believe this is workaround because of the library , right ?. 

If so, I suggest to include here a comment to explain it, referencing the library "??linked-header-??" and that it is expected to be removed if library is fixed.

blcham commented on vite.config.js:
```diff
@@ -22,6 +22,6 @@ export default defineConfig({
   },
   define: {
-    'process.env': process.env
+    "process.env": process.env, // temporary workaround for parse-link-header process.env issue
```
```suggestion
    "process.env": process.env, // workaround for parse-link-header library that depends on 2 vars defined in `process.env`, see https://github.com/thlorenz/parse-link-header/issues/31
```

blcham replied:
The motivation for the comment is to be able to easily check that the issue was solved. If there is no precise description or a link here, nobody will check it later.

blcham commented on vite.config.js:
```diff
@@ -0,0 +1,27 @@
+    port: 3000,
+  },
+  build: {
+    outDir: 'build',
```
We want to go with a default setting of Vite as much as possible. 

On the other hand, we can modify Dockerfile to achieve the same.

shellyear replied:
Ok, I set the outdir explicitly as 'dist', and changed the "build" to the "dist" in paths in the Dockerfile

blcham replied:
i would go with default value as in fta-fmea-ui -- https://github.com/kbss-cvut/fta-fmea-ui/blob/e1a243c59f12c55fae8fe66ad8b558bd104b0263/vite.config.js#L9. Or do you see an argument to have it explicit?

shellyear replied:
I wanted to make it explicit, so that it envDir will not be accidentally set to another value later. But I can remove it, if it is not necessary

blcham replied:
believe you meant `outDir` instead of `envDir`. To me, if anybody changes the `outDir`,  we will immediately find out in GitHub action so that is why I do not care to have it explicit and instead keep the configuration of Vite simpler + same everywhere. Of course, this is almost a matter of taste.

blcham commented on vite.config.js:
```diff
@@ -0,0 +1,27 @@
+  envPrefix: "RECORD_MANAGER_",
+  plugins: [react(), tsconfigPaths()],
+  server: {
+    port: 3000,
```
Instead of this modify Dockerfile to include cors origin.

blcham replied:
sorry i meant docker-compose.yaml not Dockerfile

blcham commented on vite.config.js:
```diff
@@ -0,0 +1,27 @@
+    port: 3000,
+  },
+  preview: {
+    port: 3000,
```
Instead of this modify Dockerfile to include origin.

blcham replied:
sorry i meant docker-compose.yaml not Dockerfile

blcham commented on .env.development:
```diff
@@ -0,0 +1,11 @@
+RECORD_MANAGER_API_URL=http://localhost:1235/record-manager/services/record-manager-server
+RECORD_MANAGER_APP_TITLE=Record Manager
+RECORD_MANAGER_DEV_SERVER_PORT=3000
```
Is this really in this set up configurable? Rather remove if not.

blcham commented on .env.development:
```diff
@@ -0,0 +1,11 @@
+RECORD_MANAGER_API_URL=http://localhost:1235/record-manager/services/record-manager-server
+RECORD_MANAGER_APP_TITLE=Record Manager
+RECORD_MANAGER_DEV_SERVER_PORT=3000
+RECORD_MANAGER_PROD_SERVER_PORT=8080
```
see comment above.

blcham commented on .env.production:
```diff
@@ -0,0 +1,11 @@
+RECORD_MANAGER_API_URL=http://localhost:1235/record-manager/services/record-manager-server
+RECORD_MANAGER_APP_TITLE=Record Manager
+RECORD_MANAGER_DEV_SERVER_PORT=3000
```
see comment above.

blcham commented on .env.production:
```diff
@@ -0,0 +1,11 @@
+RECORD_MANAGER_API_URL=http://localhost:1235/record-manager/services/record-manager-server
+RECORD_MANAGER_APP_TITLE=Record Manager
+RECORD_MANAGER_DEV_SERVER_PORT=3000
+RECORD_MANAGER_PROD_SERVER_PORT=8080
```
see comment above.

blcham commented on index.html:
```diff
@@ -0,0 +1,33 @@
+      integrity="sha384-Vkoo8x4CGsO3+Hhxv8T/Q5PaXtkKtu6ug5TOeNV6gBiFeWPGFN9MuhOf23Q9Ifjh"
+      crossorigin="anonymous"
+    />
+    <link rel="icon" href="/favicon.ico" />
```
Why this starts with / and config.js is relative path ? is it ok?

blcham commented on index.html:
```diff
@@ -0,0 +1,33 @@
+      name="viewport"
+      content="width=device-width, initial-scale=1, user-scalable=0"
+    />
+    <script type="text/javascript" src="locale/cs.js"></script>
```
I do not understand why here is cs.js but not en.js

Either include both of them or none of them.

blcham commented on index.html:
```diff
@@ -0,0 +1,33 @@
+      content="width=device-width, initial-scale=1, user-scalable=0"
+    />
+    <script type="text/javascript" src="locale/cs.js"></script>
+    <script src="https://unpkg.com/bowser@2.7.0/es5.js"></script>
```
do we need it?

blcham commented on index.html:
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
```
Can we somehow simplify this html file, maybe having something similar like:
https://github.com/kbss-cvut/fta-fmea-ui/blob/e1a243c59f12c55fae8fe66ad8b558bd104b0263/index.html

blcham commented on package.json:
```diff
@@ -31,57 +33,29 @@
     "redux-devtools-extension": "^2.13.9",
     "redux-thunk": "^2.3.0",
-    "util": "^0.12.5"
+    "util": "^0.12.5",
```
Why do we need this?

blcham reviewed (approved):
LGTM, but please test it in the dockerized version

blcham reviewed (changes requested):
The last changes that needs to be implemented:
- make sure that:
    - unsuported browsers are checked and message is shown to user
    - make sure that appInfo and version of the application is shown

Make sure that it is tested using docker-compose.
 
![image](https://github.com/akaene/record-manager-ui/assets/3962237/d4fc516d-e991-43ef-ad5e-1e3512a59bcd)

shellyear committed:
[feature] migrate webpack to vite;

remove everything webpack related
add vite config
format code and docker-compose to be able to work with vite
add env variables for dev and prod
remove port env varibles from .env.development and .env.production
install and use bootstrap properly
remove webpack-related package
remove unnecessary vite ts package
reimplement footer
reimplement unsupported browsers notice
update @kbss-cvut/s-forms package

blcham commented on index.html:
```diff
@@ -1,46 +1,16 @@
-    </div>
-</footer>
-<script>
-    var unsupportedBrowserElement,SUPPORTED_BROWSERS={edge:">=14",chrome:">=48",chromium:">=48",opera:">=35",safari:">=10.1",firefox:">=44",ie:">=11",samsung_internet:">=6.2"},browser=bowser.getParser(window.navigator.userAgent),isSupported=browser.satisfies(SUPPORTED_BROWSERS);
```
We lost this feature.

blcham commented on index.html:
```diff
@@ -1,46 +1,16 @@
-<footer class="footer">
-    <div class="container-fluid">
-        <p class="text-muted" style="float: left">
-            <%=htmlWebpackPlugin.options.appInfo %>
```
We lost this feature.

blcham commented on index.html:
```diff
@@ -1,46 +1,16 @@
-            <%=htmlWebpackPlugin.options.appInfo %>
-        </p>
-        <p class="text-muted" style="float: right">
-            Version: <%=htmlWebpackPlugin.options.version %>
```
We lost this feature.

blcham commented:
@blcham note we can simplify Dockerfile after this is merged.

shellyear commented:
> The last changes that needs to be implemented:
> 
> * make sure that:
>   
>   * unsuported browsers are checked and message is shown to user
>   * make sure that appInfo and version of the application is shown
> 
> Make sure that it is tested using docker-compose.
> 
> ![image](https://private-user-images.githubusercontent.com/3962237/307968981-d4fc516d-e991-43ef-ad5e-1e3512a59bcd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMjY4MjgsIm5iZiI6MTcwOTEyNjUyOCwicGF0aCI6Ii8zOTYyMjM3LzMwNzk2ODk4MS1kNGZjNTE2ZC1lOTkxLTQzZWYtYWQ1ZS0xZTM1MTJhNTliY2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjhUMTMyMjA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Zjc3ZDRhNjZlYTRlMjMwNzIzNGNhYmFjMThmMGUwNjg3NjdhZDU5ZTU1YWQzODBkOWIyNWU3NGZjYjI5YzU2ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.E0PO3kZIx-1B6I8X5oPDQJaBgDYI8hCZvpAQuo22sRg)

UPD: solved in [[5bf663d860df258661ebec5c3a277518177f7387]](https://github.com/akaene/record-manager-ui/pull/4/commits/21048153768a8c1fb374c91d900edec9d211aa86)

blcham commented on .env.development:
```diff
@@ -0,0 +1,11 @@
+RECORD_MANAGER_API_URL=http://localhost:1235/record-manager/services/record-manager-server
+RECORD_MANAGER_APP_TITLE=Record Manager
+RECORD_MANAGER_APP_INFO="blascom@gmail.com"
+RECORD_MANAGER_VERSION="0.2.3"
```
Toto je prece build time informace vychazejici z package.json. Neni dobry napad to konfigurovat, protoze to budes muset vzdycky menit ... neda se toto vymyslet nejak automaticky jako to bylo ve webpacku?

shellyear replied:
Smazal jsem RECORD_MANAGER_VERSION="0.2.3" v envu, a pouzil jsem tuto hodnou v kodu rovnou takhle "packageJson.version"

blcham replied:
OK, super, jenom zkontroluj ze funguje i v dockerizaci.

blcham commented on src/utils/BrowserCompatibilityModal.jsx:
```diff
@@ -0,0 +1,49 @@
+import { Modal } from "react-bootstrap";
+import { createPortal } from "react-dom";
+
+const SUPPORTED_BROWSERS = {
```
I believe those numbers are not correct. They were related to webpack.

blcham replied:
Just curious, there is no plugin to do that automatically, right?

shellyear replied:
We had a bowser package that was imported directly in the index.html, but I deleted it, thinking, that it was webpack-related package, but in fact it helps to check if the browsers are supported, which we can trace looking at the script for the "unsupported browsers" in index.html. So I took logic from that script and reimplemented it in the BrowserCompatibilityModal.jsx.  [bowser](https://www.npmjs.com/package/bowser)

blcham replied:
Understood, but I am only questioning whether those versions are ok even for Vite:
```  
 edge: ">=14",
  chrome: ">=48",
  chromium: ">=48",
  opera: ">=35",
  safari: ">=10.1",
  firefox: ">=44",
  ie: ">=11",
  samsung_internet: ">=6.2",
```

Because when I go to webpage https://vitejs.dev/guide/build#browser-compatibility, it shows me:
```
Chrome >=87
Firefox >=78
Safari >=14
Edge >=88
```

So my question is following: Will it really work for every version from `chrome: ">=48"` ? Or we need to update it somehow ?

shellyear replied:
You're right, those browser version listed in the link you've provided are valid, and it is written in docs "The production bundle assumes support for modern JavaScript. By default, Vite targets browsers which support the [native ES Modules](https://caniuse.com/es6-module), [native ESM dynamic import](https://caniuse.com/es6-module-dynamic-import), and [import.meta](https://caniuse.com/mdn-javascript_operators_import_meta):"

So I will change versions of the browsers as provided in the docs, but there are some other browsers such as opera, samsung_internet, chromium, that are not listed in the docs, and I'm not now sure what to do with them.
But at the same time we can use polyfills for the outdated browsers, as described here https://vitejs.dev/guide/build#browser-compatibility 

https://github.com/vitejs/vite/tree/main/packages/plugin-legacy

blcham replied:
I would:
- list only supported browsers from the documentation
- provide no polyfills (maybe we revise this decision later)

blcham commented on .env.production:
```diff
@@ -0,0 +1,11 @@
+RECORD_MANAGER_API_URL=http://localhost:1235/record-manager/services/record-manager-server
+RECORD_MANAGER_APP_TITLE=Record Manager
+RECORD_MANAGER_APP_INFO="blascom@gmail.com"
```
Tady potrebuji mit flexibilitu dat cokoli proto ne EMAIL_INFO ale APP_INFO, priklady:

1) viz Dockerfile field default value:
`ENV APP_INFO='<a href="https://github.com/blcham" target="_blank" rel="noopener noreferrer" title="github.com/blcham, 2023">©&nbsp;github.com/blcham, 2023</a>'`

2) viz. obrazek "For technical help contact blaskom@gmail.com"

3) © KBSS at FEE CTU in Prague, 2024

By default bych to nechal prazdny.

shellyear replied:
Neni dobry vnorovat achor tady do html, protoze to budu muset udelat pres dangerouslySetInnerHtml (coz neni bezpecny, protoze je tam riziko XSS utoku). 
Pokud to nechame prazdny, tak bych mel tenhle link nadefinovat v kodu natvrdo ? 
Kde ve footeru bych mel rozmistit "© KBSS at FEE CTU in Prague, 2024" ?
Aktualne footer vypada takhle 
<img width="1439" alt="Screenshot 2024-02-28 at 17 44 23" src="https://github.com/akaene/record-manager-ui/assets/50382323/647f9517-54b6-4bf1-b047-b7b79e3da831">

blcham replied:
OK, good point !, I suggest putting there only text defined in APP_INFO

blcham replied:
So the text could be "© KBSS at FEE CTU in Prague, 2024". Leave default blank.

shellyear replied:
So there will be "version" on the right side of the footer. And the left side will only have  "© KBSS at FEE CTU in Prague, 2024" which is a value extracted from RECORD_MANAGER_APP_INFO and wrapped in a link with href="https://github.com/blcham", right ?

blcham replied:
I suggest two alternative solutions (Implement whichever you like) :
1) do not include link at all
2) include link in the environment variable  RECORD_MANAGER_APP_INFO_LINK. It may also be empty which would cause no <a href ..> being rendered.

Of course I prefer 2), but have no problem with 1) at all.

shellyear replied:
I would like to save the _RECORD_MANAGER_APP_INFO_LINK_ with the link, to use it in the dockerFile, and use some other env variable for the client code(e.g just _RECORD_MANAGER_APP_INFO_), which will not contain any html tags, but will contain the actual link, and then in the client code in _JSX_ I will have an _achor_ tag, and will just use that env variable as _href_. What do you think ? 

Otherwise I would like to avoid using dangerouslySetInnerHtml, but if for now it is more important, we can continue with the 2 option

shellyear replied:
Actual state is on the image, I used RECORD_MANAGER_APP_INFO, but it is empty in .env file, then in config/index.js I set the default value for this variable with the given text. Also, I've squashed last couple of commits(that solve this). Please let me know if it is ok 
<img width="862" alt="Screenshot 2024-02-29 at 10 21 59" src="https://github.com/akaene/record-manager-ui/assets/50382323/0adbe1c7-49c9-4a8c-9f23-93b10e5b6dab">

blcham reviewed (changes requested):
see my comments.

blcham commented on src/styles/record-manager.css:
```diff
@@ -237,7 +249,7 @@ td.actions button {
 @media screen and (min-width: 576px) and (max-width: 992px) {
     .container {
-        width: auto;
+        width: 100%;
```
why this? Moreover, please test it on different resolutions.

shellyear replied:
@blcham I've tested it again, and decided to have width of the container as 100% on this resolution, because flex items shrink, when the **width** is **auto**. (pic)
Also more CSS needs to be changed in my opinion, to make it more simple, but I would like to do it in another branch

<img width="499" alt="Screenshot 2024-03-01 at 14 05 37" src="https://github.com/akaene/record-manager-ui/assets/50382323/0b983463-d03a-4bf3-8cac-428359eded17">

shellyear commented:
@blcham I've found a new error, when I go to the "/records". It seems that **@kbss-cvut/s-forms** uses outdated **util.isArray** function, and the **Array.isArray** should be used instead,  https://github.com/mysticatea/eslint-plugin-node/issues/65. I think to solve it directly in that package. 

Every other route works, but also I don't have any available options to create a record(last img)

UPD: I have also removed unnecessary CSS (a little), and squashed the commits

<img width="906" alt="Screenshot 2024-02-29 at 18 06 48" src="https://github.com/akaene/record-manager-ui/assets/50382323/05f87c63-f24c-416a-9458-c656579a3a6e">
<img width="903" alt="Screenshot 2024-02-29 at 18 06 37" src="https://github.com/akaene/record-manager-ui/assets/50382323/9435c186-8d41-48ab-937b-c6b120524b5a">
<img width="910" alt="Screenshot 2024-02-29 at 18 08 55" src="https://github.com/akaene/record-manager-ui/assets/50382323/4fa12c48-a294-465e-b81c-f6ee4599a6ac">

blcham commented:
> @blcham I've found a new error, when I go to the "/records". It seems that **@kbss-cvut/s-forms** uses outdated **util.isArray** function, and the **Array.isArray** should be used instead, [mysticatea/eslint-plugin-node#65](https://github.com/mysticatea/eslint-plugin-node/issues/65). I think to solve it directly in that package.

@shellyear  Thanks! It was a dependent library [tsort](https://www.npmjs.com/package/tsort) which was last updated 11 years ago :), so I had to fork the library and create [a new version](https://www.npmjs.com/package/@kbss-cvut/tsort). The fix is in commit https://github.com/kbss-cvut/s-forms/commit/0030c82472987079169abb4d5c18233d2c957f9e and you should fix the issue by upgrading the dependency:

`npm i @kbss-cvut/s-forms@0.7.1-beta-48c6cc4.0`

shellyear commented:
> @shellyear Thanks! It was a dependent library [tsort](https://www.npmjs.com/package/tsort) which was last updated 11 years ago :), so I had to fork the library and create [a new version](https://www.npmjs.com/package/@kbss-cvut/tsort). The fix is in commit [kbss-cvut/s-forms@0030c82](https://github.com/kbss-cvut/s-forms/commit/0030c82472987079169abb4d5c18233d2c957f9e) and you should fix the issue by upgrading the dependency:
> 
> `npm i @kbss-cvut/s-forms@0.7.1-beta-48c6cc4.0`

Nice, thanks. Just upgraded the package

shellyear commented:
Created an issue about CSS refactorings [here](https://github.com/kbss-cvut/record-manager-ui/issues/104)
